text/0066-separate-context-propagation.md

```diff
@@ -0,0 +1,557 @@
+blocks.
+
+For backwards compatibility, OpenTracing Baggage is propagated as Correlations 
+when using the OpenTracing bridge.
+
+As every correlation adds additional overhead to every request, correlations 
+should be used with care. The implementation MUST propagate these values. It is 
+entirely on the user to choose when to drop correlations, either by removing 
+them via the API, or by filtering them via a network proxy or service mesh.
+
+**`GetCorrelation(context, key) -> value`**  
+To access the value for a label set by a prior event, the Correlations API 
+provides a function which takes a context and a key as input, and returns a value.
+
+**`SetCorrelation(context, key, value) -> context`**  
+To record the value for a label, the Correlations API provides a function which 
+takes a context, a key, and a value as input, and returns an updated context 
+which contains the new value.
+
+**`RemoveCorrelation(context, key) -> context`**  
+To delete a label, the Correlations API provides a function 
+which takes a context and a key as input, and returns an updated context which 
+no longer contains the selected key-value pair.
+
+**`ClearCorrelations(context) -> context`**  
+To avoid sending any labels to an untrusted process, the Correlation API 
+provides a function to remove all Correlations from a context.
+
+**`GetCorrelationPropagator() -> (HTTP_Extractor, HTTP_Injector)`**  
+To deserialize the previous labels set by prior processes, and to serialize the 
+current total set of labels and send them to the next process, the Correlations 
+API provides a function which returns a Correlation-specific implementation of 
+the `HTTPExtract` and `HTTPInject` functions found in the Propagation API.
+
+# Context Propagation
+
+## Context API
+
+Cross-cutting concerns access data in-process using the same, shared context 
+object. Each concern uses its own namespaced set of keys in the context, 
+containing all of the data for that cross-cutting concern.
+
+**`GetValue(context, key) -> value`**  
+To access the local state of an concern, the Context API provides a function 
+which takes a context and a key as input, and returns a value.
+
+**`SetValue(context, key, value) -> context`**  
+To record the local state of a cross-cutting concern, the Context API provides a 
+function which takes a context, a key, and a value as input, and returns an 
+updated context which contains the new value.
+
+
+### Optional: Automated Context Management
+When possible, the OpenTelemetry context should automatically be associated 
+with the program execution context. Note that some languages do not provide any 
+facility for setting and getting a current context. In these cases, the user is 
+responsible for managing the current context.  
+
+**`GetCurrent() -> context`**  
+To access the context associated with program execution, the Context API 
+provides a function which takes no arguments and returns a Context.
+
+**`SetCurrent(context)`**  
```

Comment:
SetCurrent should return a token that is used to restore. See io.grpc.Context attach/detach.

Reply:
Since context methods using builder pattern - in order to add a single value to the context you need to do:

```
SetCurrrent(GetCurrent().SetCorrelation(key, value)).
```

Is it correct understanding?

Reply:
@bogdandrutu I agree, this API missing something like a closure. In some languages this is fine (Erlang) but in most languages it would look more like:

```
SetScope(context, closure)
```

@SergeyKanzhelev it would work similar to what you describe. Except that if you are using automated context propagation, you can remove the explicit context from the function signatures, and have the context reference be managed automatically for you.

```
// automatic context handling
SetCorrelation(key, value)

// the explicit form is equivalent, but not normally needed
SetContext( SetCorrelation( GetContext(), key, value))
```

IMHO, this is the trickiest piece to spec, since every language does it differently and has different facilities. Java and .NET is on one end, Erlang and Go are on the other. 

Here are the related examples (which would need to be updated in tandem):
* https://github.com/tedsuo/rfcs/blob/context-prop-2/text/0066-separate-context-propagation.md#simplify-the-api-with-automated-context-propagation
* https://github.com/tedsuo/rfcs/blob/context-prop-2/text/0066-separate-context-propagation.md#the-scope-of-current-context